### PR TITLE
Megafauna

### PR DESCRIPTION
**Pull request by imtakingabreakdontatme**
Very large monsters that will appear rarely in lavaland. They are intended for large parties equipped with RD nonsense to kill. @TechnoAlchemisto asked for monster hunter so blame him.

Legion floats around spitting the smaller legion heads, periodically charging forwards rapidly to attack. Killing legion splits it into two smaller legions, and this goes on for quite some time. The final legion head will drop the Staff of Storms .Sprite is from @Cheridan. Maybe one day it can have more interesting loot, but I doubt anyone is going to kill it under normal circumstances anyway.

Ash Drakes are more minibosses/large mobs, they are doable with a coordinated team of miners (or maybe even solo). They breathe fire, swoop into the air/divebomb, and rain fire from the sky. They drop a special chest which has three new items (which are mostly placeholders until I get around to making dragon related crafting materials).

:cl: Kor
rscadd: Megafauna have been added to lavaland.
rscadd: Ash Drakes have been spotted roaming the wastes. Use your GPS to track their fiery signals.
rscadd: You can now knock on the Necropolis door, but it's probably best that you don't.
/:cl:

### CODE REVIEW

**Comment by Cheridan** [comment]
>dragon

Ridley is too big for Brawl but I guess he's fine in lavaland

**Comment by KazeEspada** [comment]
Can we have a large parrot like monster?
https://www.google.com/url?sa=i&rct=j&q=&esrc=s&source=images&cd=&ved=0ahUKEwjbxrXw757MAhWD1CYKHb-UBp4QjRwIBw&url=http%3A%2F%2Fmonsterhunter.wikia.com%2Fwiki%2FYian_Kut-Ku_Photo_Gallery&psig=AFQjCNHsC_VKzcSHAWq7scKsyUfiUmVTuQ&ust=1461298662558957

**Comment by TechnoAlchemisto** [comment]
:hooray:

**Comment by RemieRichards** [comment]
I'd prefer if the actual boss framework was brought over with this (I know I've been lazy with reviving that PR)

**Comment by KorPhaeron** [comment]
@RemieRichards 

You mean like the anger stuff? I was gonna do that for the dragon, but it's not needed for the legion

**Comment by RemieRichards** [comment]
@KorPhaeron the whole ATB+anger thing ye, I was supposed to generalise it to a ```/boss``` class but I guess I never did.

**Comment by KorPhaeron** [comment]
@RemieRichards 

I actually did add some of that anger stuff as the boss class (increasing anger as you get hit, the anger modifier proc), but I thought the actual AI/how they interpreted the anger should be left up to each individual boss

**Comment by RemieRichards** [comment]
@KorPhaeron my idea was an ATB system of ability points (that they can spend on ability datums) that slowly recharge, boss AI would be ```handle_boss_ai()``` which just hooks into ```handle_automated_action()```

File: code/modules/mob/living/simple_animal/hostile/megafauna/megafauna.dm [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	else
+		..()
+
+/mob/living/simple_animal/hostile/megafauna/adjustBruteLoss(damage)
+	..()
+	anger += damage
+
+/mob/living/simple_animal/hostile/megafauna/proc/anger_modifier()
+	. = 0
+	if(anger)
```
Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/hostile/megafauna/megafauna.dm]: this entire proc isn't functional you know.

Comment (KorPhaeron) [inline code/modules/mob/living/simple_animal/hostile/megafauna/megafauna.dm]: not surprising

Comment (KorPhaeron) [inline code/modules/mob/living/simple_animal/hostile/megafauna/megafauna.dm]: Maybe I should just do this PR as is (with the death overrides, and Legion finished) and do the dragon later after you've had a chance to add your AI to it

**Comment by Bawhoppen** [comment]
kor are you sure that dragon sprite isn't from dragon cave

**Comment by Robustin** [comment]
If we're going down the MMORPG road with this, might as well add a staple of MMOs:

AoE indicators for certain attacks, so you can have certain fight mechanics that are difficult, but fair, to manage.

**Comment by FantasticFwoosh** [comment]
Will any of these boss monsters be interactable with xenobiology as a planned move? (as per sentience/creation/transferrence to use special boss abilities)

Dragonfire lava heat & magic-karp fireball combustion heat exchange wizard engine is go.

File: code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+/mob/living/simple_animal/hostile/megafauna/legion/death()
+	if(size > 0)
+		for(var/i in 1 to 2)
+			var/mob/living/simple_animal/hostile/megafauna/legion/L = new(src.loc)
+			L.resize = src.resize * 0.8
+			L.update_transform()
+			L.health = src.maxHealth * 0.8
+			L.maxHealth = L.health
+			L.size--
```
Comment (ChangelingRain) [inline code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm]: Shouldn't this be L.size = size--

**Comment by RandomMarine** [comment]
Big monsters with extremely high health pools need to react to damage, otherwise players will question whether their attacks aren't doing anything at all. Anger stuff is one thing, but simply making sounds or even going so far as to causing occasional stuns/knockdowns are a good morale boost to players.

File: code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm (hunk 2 above)
Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm]: ```L.size = --size```

**Comment by KorPhaeron** [comment]
I forgot how weird big mobs work with projectiles.

**Comment by GunHog** [comment]
> Big monsters with extremely high health pools need to react to damage, otherwise players will question whether their attacks aren't doing anything at all. Anger stuff is one thing, but simply making sounds or even going so far as to causing occasional stuns/knockdowns are a good morale boost to players.

You could add a health bar! That is how some other games handled tanky mobs. Eh, but then you would need a new HUD or you would have to give miners Medical HUDs.

Oh, another idea - damage states. The dragon's hide shows cracks and scales fall off as damage progresses, similar to how synths lose their skin as they take damage!

**Comment by Iamgoofball** [comment]
We could also do hit numbers

**Comment by drynwyn** [comment]
They should have sounds or something as you get near them- otherwise all the miners who don't know how to use the GPS will die- 

Nevermind fine as is.

**Comment by KorPhaeron** [comment]
@RemieRichards 

Do you have any ideas regarding hitboxes, or should I just let projectiles be weird with them? (At least clicking on them at any point sends the projectile to the right part of them)

**Comment by RandomMarine** [comment]
@KorPhaeron Give the sprites an obvious-looking weak point that matches up with which tile the mob is on.

File: code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	if(world.time >= ranged_cooldown && !charging)
+		if(prob(75))
+			var/mob/living/simple_animal/hostile/asteroid/hivelordbrood/legion/A = new(src.loc)
+			A.GiveTarget(target)
+			A.friends = friends
+			A.faction = faction
+			ranged_cooldown = world.time + ranged_cooldown_time
+		else
+			visible_message("<span class='danger'>[src] charges!</span>")
+			SpinAnimation(speed = 20, loops = 5)
```
Comment (KazeEspada) [inline code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm]: This looks terrible in game TBH

File: code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	else
+		var/last_legion = TRUE
+		for(var/mob/living/simple_animal/hostile/megafauna/legion/other in mob_list)
+			if(other != src)
+				last_legion = FALSE
+				break
+		if(last_legion)
+			src.loot = list(/obj/item/weapon/staff_of_storms)
+		else if(prob(5))
+			src.loot = list(/obj/structure/closet/crate/necropolis)
```
Comment (ChangelingRain) [inline code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm]: The loot is already set to be this, so the prob() does nothing.

File: code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+	gpstag = "Echoing Signal"
+	desc = "The message repeats."
+	invisibility = 100
+
+
+//Loot
+
+/obj/item/weapon/staff_of_storms
+	name = "staff of storms"
+	desc = "An ancienct staff retrieved from the remains of Legion. The wind stirs as you move it."
```
Comment (ChangelingRain) [inline code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm]: ancient

File: code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm (hunk 4 above)
Comment (KorPhaeron) [inline code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm]: Oops, I meant to make the standard loot bones

**Comment by KorPhaeron** [comment]
I need to consider how/where it spawns and its aggro range, would suck if someone just pulled this back to the mining base (as I saw happen with a free golem colony)

**Comment by RandomMarine** [comment]
Would it be possible for it to check its map coordinates, like, if it goes further south than 70 Y, it warps back to its initial spawnpoint?

**Comment by KorPhaeron** [comment]
that'd be doable yeah, might lead to weird shit like constantly shooting it and running back over the line though.

File: code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+	damtype = BURN
+	hitsound = 'sound/weapons/sear.ogg'
+	var/storm_cooldown = 0
+
+/obj/item/weapon/staff_of_storms/attack_self(mob/user)
+	if(storm_cooldown > world.time)
+		user << "The staff is still recharging."
+		return
+
+	var/obj/machinery/lavaland_controller/linked_machine
```
Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm]: put this on the staff, and only find a new one if it's become null, that way this loop isn't done needlessly.

Comment (KorPhaeron) [inline code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm]: For some reason I thought to myself "this should work on every z level that has weather" despite the chances of more weather z levels being basically zero

Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm]: you can do that too, just make it ```if(!linked_machine || linked_machine.z != user.z)``` silly.

**Comment by RandomMarine** [comment]
I think it's about time to separate lavaland into more areas, then.

One for the furthest south of it all, where mobs can't spawn.
One intermediary area, which acts as usual. (Bonus, make mining GPSes a child type that gives a warning beep the first time they enter this area. To warn newbie miners.)
Then the whole upper half of the map can be a third area, which adds the possibility of megafauna.

**Comment by Man-Shroom-** [comment]
or multiple z-levels

Chasms drop you to the secondary z-level where the ash walker society is(maybe) and there's tons of goodies down there, intermediate level basically and there's no chasms just lava

Necro-door takes you to the forbidden third layer where the MegaFauna are and there are no chasms here just lava lava

**Comment by Dorsisdwarf** [comment]
More lavaland Z-levels sounds like fun. You can even have a baby one with low loot levels & greatly reduced spawns for drone miners/noobs

**Comment by KorPhaeron** [comment]
@Cheridan 

Okay it's sort of ready in that the dragon is finished, but none of the loot is tested, and I'd rather make crafting out of dragon shit than have it just drop weird magic items.

So it's ready to merge probably but one day I will surely give it a corpse and better loot.

**Comment by KorPhaeron** [comment]
Okay I lied I broke something and I must fix it

**Comment by GunHog** [comment]
Pics?

**Comment by KorPhaeron** [comment]
Okay it's fixed

**Comment by GunHog** [comment]
dragon

**Comment by KorPhaeron** [comment]
Its negatively impacted the ghost roles when they get wiped out every round by a roving boss

**Comment by GunHog** [comment]
As in it randomly swoops down and cleans them out?

**Comment by KorPhaeron** [comment]
Someone will stumble into a boss, they'll panic and run back to the base, the lizards armed with spears will get wrecked

**Comment by TechnoAlchemisto** [comment]
lizards dying sounds like it's working as intended to be honest

**Comment by GunHog** [comment]
Sounds like it is their own fault. If you upset a **GIGANTIC DRAGON,** the last thing you should do is lead it back to your house!

**Comment by KorPhaeron** [comment]
What about the poor golems though?

**Comment by PKPenguin321** [comment]
It will be a mighty war

**Comment by Iamgoofball** [comment]
@KorPhaeron sounds like a good penalty for signing up as a ghost role instead of a station role

**Comment by GunHog** [comment]
I do not like to look it at as a "penalty". The ghost roles taken are used by more than just the players who decided against joining the station. There are those that take a ghost role after dying on the station. It happens quite often, you know. 

Besides, there are good reasons why some choose to observe rather than join:

- Perhaps they did not intend on playing for a long time - most jobs require a longer commitment into the round. 
- The roles a player prefers are taken, and he would rather not be forced to play as something else.
- Unstable or unreliable internet connection.
- They simply do not want to play, but only watch. Admins do not care if a player abandons a ghost role in most cases.
- Ghost roles are sometimes more enjoyable than station ones.

**Comment by KorPhaeron** [comment]
Okay I just threw Legion behind the door, and you can go knock on it if you're feeling crazy.

@Cheridan 

This is ready to merge now.

**Pull request merged**